bigtable/metrics.go

```diff
@@ -0,0 +1,426 @@
+)
+
+var (
+	// duration between two metric exports
+	// Not a constant. Modifiable in tests
+	defaultSamplePeriod = 5 * time.Minute
+
+	clientName = fmt.Sprintf("go-bigtable/%v", internal.Version)
+
+	// Arrays cannot be constants in Go
+	bucketBounds = []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0, 10.0, 13.0, 16.0, 20.0, 25.0, 30.0, 40.0,
+		50.0, 65.0, 80.0, 100.0, 130.0, 160.0, 200.0, 250.0, 300.0, 400.0, 500.0, 650.0,
+		800.0, 1000.0, 2000.0, 5000.0, 10000.0, 20000.0, 50000.0, 100000.0, 200000.0,
+		400000.0, 800000.0, 1600000.0, 3200000.0}
+
+	// All the built-in metrics have same attributes except 'status' and 'streaming'
+	// These attributes need to be added to only few of the metrics
+	// Maps cannot be constants in Go
+	additionalAttributes = map[string][]string{
+		metricNameOperationLatencies: {
+			metricLabelKeyStatus,
+			metricLabelKeyStreamingOperation,
+		},
+		metricNameAttemptLatencies: {
+			metricLabelKeyStatus,
+			metricLabelKeyStreamingOperation,
+		},
+		metricNameServerLatencies: {
+			metricLabelKeyStatus,
+			metricLabelKeyStreamingOperation,
+		},
+		metricNameRetryCount: {
+			metricLabelKeyStatus,
+		},
+	}
+
+	// Generates unique client ID in the format go-<random UUID>@<>hostname
+	generateClientUID = func() (string, error) {
+		hostname := "localhost"
+		hostname, err := os.Hostname()
+		if err != nil {
+			return "", err
+		}
+		return "go-" + uuid.NewString() + "@" + hostname, nil
+	}
+
+	// Not a constant. Modifiable in tests
+	exporterOpts = []option.ClientOption{}
+)
+
+type builtinMetricsTracerFactory struct {
+	builtinEnabled bool
+
+	// To be called on client close
+	shutdown func()
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	operationLatencies metric.Float64Histogram
+	serverLatencies    metric.Float64Histogram
+	attemptLatencies   metric.Float64Histogram
+	retryCount         metric.Int64Counter
+}
+
+func newBuiltinMetricsTracerFactory(ctx context.Context, project, instance, appProfile string, metricsProvider MetricsProvider) (*builtinMetricsTracerFactory, error) {
+	clientUID, err := generateClientUID()
+	if err != nil {
+		log.Printf("built-in metrics: generateClientUID failed: %v. Using empty string in the %v metric atteribute", err, metricLabelKeyClientUID)
+	}
+
+	tracerFactory := &builtinMetricsTracerFactory{
+		builtinEnabled: false,
+		clientAttributes: []attribute.KeyValue{
+			attribute.String(monitoredResLabelKeyProject, project),
+			attribute.String(monitoredResLabelKeyInstance, instance),
+			attribute.String(metricLabelKeyAppProfile, appProfile),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+		},
+		shutdown: func() {},
+	}
+
+	if emulatorAddr := os.Getenv("BIGTABLE_EMULATOR_HOST"); emulatorAddr != "" {
+		// Do not emit metrics when emulator is being used
+		return tracerFactory, nil
+	}
```

Comment:
please centralize the references to global env vars ... this should only be pulled from the env by the client and then passed around as an actual parameter. Ideally the client would explicitly set the noop tracer factory instead of bleeding the logic into the tracerfactory

Reply:
Done